torch/tensor.py

```diff
@@ -441,4 +441,52 @@ def __array_wrap__(self, array):
             array = array.astype('uint8')
         return torch.from_numpy(array)
 
+    @property
+    def __cuda_array_interface__(self):
+        """Array view description for cuda tensors.
+
+        See:
+        https://numba.pydata.org/numba-doc/latest/cuda/cuda_array_interface.html
+        """
+
+        # raise AttributeError for unsupported tensors, so that
+        # hasattr(cpu_tensor, "__cuda_array_interface__") is False.
+        if not self.device.type == "cuda":
+            raise AttributeError(
+                "Not on cuda device, use Tensor.cuda() first: %r" %
+                self.device
+            )
+
+        if self.is_sparse:
+            raise AttributeError(
+                "Can't convert sparse tensor, use Tensor.to_dense() "
+                "to convert to a dense tensor first."
+            )
+
+        # RuntimeError, matching tensor.__array__() behavior.
+        if self.requires_grad:
+            raise RuntimeError(
+                "Can't get __cuda_array_interface__ on Variable that requires grad. "
+                "Use var.detach().__cuda_array_interface__ instead."
```

Comment:
If our user blindly follows this advice without knowing what detach() actually does, they might come calling us when their code didn't work, because this advice didn't solve the fundamental problem which is that you can't autograd past PyTorch. More robust advice would be to say, "If gradients are truly not required, you can get a Variable which doesn't require grad by saying var.detach()". And I wouldn't advise the user to use __cuda_array_interface__, because in all likelihood this call is happening in code out of their control.

Reply:
:+1: As before, this was an attempt to match `Tensor.numpy()`, but I agree with your points.

How about:

```
"Can't get __cuda_array_interface__ on Variable that requires grad. "
"If gradients are not required, use var.detach() to get Variable that does not require grad."
```

Reply:
I would probably also say to update `Tensor.numpy()` too :) SGTM

Reply:
To follow up, this standard message is used across a relatively wide collection of tensor interfaces when guarding against non-gradient-safe calls. I'm happy to do the update across the board if you'd like, but want to double check before I touch a wider chunk of the interface.

Ex: [numpy](https://github.com/asford/pytorch/blob/36019decacf872dfa48d99d990cf1a0cb9b257ef/tools/autograd/templates/python_variable_methods.cpp#L327-L328), [map_](
https://github.com/asford/pytorch/blob/36019decacf872dfa48d99d990cf1a0cb9b257ef/tools/autograd/templates/python_variable_methods.cpp#L400-L401), [map2_](https://github.com/asford/pytorch/blob/36019decacf872dfa48d99d990cf1a0cb9b257ef/tools/autograd/templates/python_variable_methods.cpp#L418-L419) et al...